app/services/facility_stats_service.rb

```diff
@@ -0,0 +1,71 @@
+class FacilityStatsService
+  attr_reader :facilities_data, :stats_by_size
+
+  def initialize(accessible_facilities:, retain_facilities:, ending_period:, rate_numerator:)
+    @facilities = accessible_facilities
+    @retain_facilities = retain_facilities
+    @ending_period = ending_period
+    @rate_numerator = rate_numerator
+    @rate_name = "#{rate_numerator}_rate"
+    @periods = ending_period.downto(5)
+    @facilities_data = {}
+    @stats_by_size = {}
+  end
+
+  def call
+    facilities.each do |facility|
+      facility_data = Reports::RegionService.new(region: facility, period: ending_period).call
+      add_facility_stats(facility_data)
+      @facilities_data[facility.name] = facility_data if retain_facility?(facility)
+    end
+    add_control_rates
+  end
+
+  private
+
+  attr_reader :facilities, :retain_facilities, :ending_period, :rate_numerator, :rate_name, :periods
+
+  def add_facility_stats(facility_data)
+    size = facility_data.region.source.facility_size
+    add_size_section(size) unless stats_by_size[size]
+    periods.each do |period|
+      current_period = stats_by_size[size][period]
+      current_period[rate_numerator] += facility_data.dig(rate_numerator, period) || 0
+      current_period["adjusted_registrations"] += facility_data["adjusted_registrations"][period]
+      current_period["cumulative_registrations"] += facility_data["cumulative_registrations"][period]
+    end
+  end
+
+  def add_control_rates
```

Comment:
Is this always control rates?  I think sometimes it can be `missed_visits` ...so maybe `calculate_percentages` ?

Reply:
This needs to be renamed. My first pass through this I was working with control rates and named some things too specifically. Thanks for catching this.